src/Microsoft.TemplateEngine.Core/Conditional.cs

```diff
@@ -206,85 +290,164 @@ public int HandleMatch(IProcessorState processor, int bufferLength, ref int curr
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
             }
 
-            private void SkipToMatchingEndif(IProcessorState processor, ref int bufferLength, ref int currentBufferPosition, ref int token)
+            // Moves the buffer to the next token at the same level of nesting as the current token.
+            // Should never be called if we're on an end token!!!
+            // Returns false if no next token can be found at the same level. 
+            //      this is probably indicative of a template authoring problem, or possibly a buffer problem.
+            private bool SeekToNextTokenAtSameLevel(IProcessorState processor, ref int bufferLength, ref int currentBufferPosition, out int token)
             {
-                while (token != 1 && SeekToTerminator(processor, ref bufferLength, ref currentBufferPosition, out token) && token != 1)
+                if (_definition.WholeLine)
                 {
-                    int balance = 1;
+                    processor.SeekForwardThrough(processor.EncodingConfig.LineEndings, ref bufferLength, ref currentBufferPosition);
+                }
 
-                    //We're in a nested if branch, wait until it balances closed
-                    while (balance > 0 && SeekToTerminator(processor, ref bufferLength, ref currentBufferPosition, out token))
+                bool seekSucceeded = SeekToToken(processor, ref bufferLength, ref currentBufferPosition, out token);
+
+                //Keep on scanning until we've hit a balancing token that belongs to us
+                // each "if" found opens a new level of nesting
+                while (IsTokenIndexOfType(token, IfTokenBaseIndex) || IsTokenIndexOfType(token, IfTokenActionableBaseIndex))
+                {
+                    int open = 1;
+                    while (open != 0)
                     {
-                        if (token == 0)
+                        seekSucceeded &= SeekToToken(processor, ref bufferLength, ref currentBufferPosition, out token);
+
+                        if (IsTokenIndexOfType(token, IfTokenBaseIndex) || IsTokenIndexOfType(token, IfTokenActionableBaseIndex))
                         {
-                            ++balance;
+                            ++open;
                         }
-                        else if (token == 1)
+                        else if (IsTokenIndexOfType(token, EndTokenBaseIndex))
                         {
-                            --balance;
+                            --open;
                         }
                     }
+
+                    seekSucceeded &= SeekToToken(processor, ref bufferLength, ref currentBufferPosition, out token);
+                }
+
+                // temporary for debugging.
+                if (! seekSucceeded)
+                {
+                    Console.WriteLine("seek to token failed in SeekToNextBalancedToken()");
```

Comment:
I’m deleting it, it was useful when making big changes to Conditional::HandleMatch(). I should have gone through this code more thoroughly before making a PR.

From: Mike Lorbetske [mailto:notifications@github.com]
Sent: Tuesday, August 23, 2016 11:12 AM
To: seancpeters/microsoft-templateengine <microsoft-templateengine@noreply.github.com>
Cc: Sean Peters <sepeters@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [seancpeters/microsoft-templateengine] Refactor (#1)


In src/Microsoft.TemplateEngine.Core/Conditional.cs<https://github.com/seancpeters/microsoft-templateengine/pull/1#discussion_r75921131>:

>                          }

>                      }

> +

> +                    seekSucceeded &= SeekToToken(processor, ref bufferLength, ref currentBufferPosition, out token);

> +                }

> +

> +                // temporary for debugging.

> +                if (! seekSucceeded)

> +                {

> +                    Console.WriteLine("seek to token failed in SeekToNextBalancedToken()");

Debug.WriteLine maybe?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/seancpeters/microsoft-templateengine/pull/1/files/a425b2f9e55c8aa446463bd90060cb6bc99db3d0#r75921131>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AT8a8YyzsgW2qI9EBY4uCMqIpsOBdvyDks5qizgEgaJpZM4JqaeQ>.